tests/rest/client/test_rooms.py

```diff
@@ -2315,3 +2315,140 @@ def test_bad_alias(self):
+        return_value = True
+
+        async def _user_may_send_3pid_invite(
+            inviter: str,
+            invitee: Dict[str, str],
+            room_id: str,
+        ) -> bool:
+            return return_value
+
+        allow_mock = Mock(side_effect=_user_may_send_3pid_invite)
+
+        self.hs.get_spam_checker()._user_may_send_3pid_invite_callbacks.append(
+            allow_mock
+        )
+
+        # Send a 3PID invite into the room and check that it succeeded.
+        email_to_invite = "teresa@example.com"
+        channel = self.make_request(
+            method="POST",
+            path="/rooms/" + self.room_id + "/invite",
+            content={
+                "id_server": "example.com",
+                "id_access_token": "sometoken",
+                "medium": "email",
+                "address": email_to_invite,
+            },
+            access_token=self.tok,
+        )
+        self.assertEquals(channel.code, 200)
+
+        # Check that the callback was called with the right params.
+        expected_call_args = (
+            (
+                self.user_id,
+                {"medium": "email", "address": email_to_invite},
+                self.room_id,
+            ),
+        )
+
+        self.assertEquals(
+            allow_mock.call_args, expected_call_args, allow_mock.call_args
+        )
+
+        # Check that the call to send the invite was made.
+        self.assertEquals(make_invite_mock.call_count, 1)
+
+        # Now change the return value of the callback to deny any invite and test that
+        # we can't send the invite.
+        return_value = False
+        channel = self.make_request(
+            method="POST",
+            path="/rooms/" + self.room_id + "/invite",
+            content={
+                "id_server": "example.com",
+                "id_access_token": "sometoken",
+                "medium": "email",
+                "address": email_to_invite,
+            },
+            access_token=self.tok,
+        )
+        self.assertEquals(channel.code, 403)
+
+        # Also check that it stopped before calling _make_and_store_3pid_invite.
+        self.assertEquals(make_invite_mock.call_count, 1)
+
+    def _mock_make_and_store_3pid_invite(self) -> Mock:
+        """Mocks RoomMemberHandler._make_and_store_3pid_invite with a function that just
+        returns the integer 0.
+
+        Returns:
+            The Mock object _make_and_store_3pid_invite was replaced with.
+        """
+
+        async def _make_and_store_3pid_invite(
+            requester: Requester,
+            id_server: str,
+            medium: str,
+            address: str,
+            room_id: str,
+            user: UserID,
+            txn_id: Optional[str],
+            id_access_token: Optional[str] = None,
+        ) -> int:
+            return 0
+
+        mock = Mock(side_effect=_make_and_store_3pid_invite)
+
+        self.hs.get_room_member_handler()._make_and_store_3pid_invite = mock
```

Comment:
If we don't care about the parameters could we just do:

```suggestion
        self.hs.get_room_member_handler()._make_and_store_3pid_invite = Mock(return_value=0)
```

A similar thing could be done for `_mock_lookup_3pid`. They could probably be inlined then instead.

Reply:
This doesn't work for all the versions of Python we support, since some of them don't support `await`ing on a `Mock` unless it has an `async` side effect

Reply:
Though now I don't remember which versions of Python I've tested this with, so I've opened https://github.com/matrix-org/synapse/pull/10928 to make the CI check if that's still the case.

Reply:
As an example, when testing your suggestion locally:

```
2021-09-27 17:13:01+0100 [-] 2021-09-27 17:13:01,823 - synapse.http.server - 93 - ERROR - POST-4 - Failed handle request via 'RoomMembershipRestServlet': <SynapseRequest at 0x7f1bb01eb0a0 method='POST' uri='/_matrix/client/r0/rooms/!vDuFLwPWZJCFsnpqsV:test/invite' clientproto='1.1' site='test'>
	Traceback (most recent call last):
	  File "/home/babolivier/Documents/matrix/synapse/synapse/http/server.py", line 258, in _async_render_wrapper
	    callback_return = await self._async_render(request)
	  File "/home/babolivier/Documents/matrix/synapse/synapse/http/server.py", line 446, in _async_render
	    callback_return = await raw_callback_return
	  File "/home/babolivier/Documents/matrix/synapse/synapse/rest/client/room.py", line 790, in on_POST
	    await self.room_member_handler.do_3pid_invite(
	  File "/home/babolivier/Documents/matrix/synapse/synapse/handlers/room_member.py", line 1278, in do_3pid_invite
	    stream_id = await self._make_and_store_3pid_invite(
	TypeError: object int can't be used in 'await' expression
```

Reply:
You can use `make_awaitable` from `tests.test_utils` for that:

`Mock(return_value=make_awaitable(0))`.

Reply:
Oooh that is true!